### PR TITLE
[native] Implement bucket conversion for Hive splits

### PR DESCRIPTION
**Pull request by Yuhta**
When the bucket count of a table changes over time, there can be legitimate cases that multiple buckets exist in the same file.  In such cases the query planner should set bucket conversion for these splits and in Velox we use extra filter to get only the rows corresponding to the bucket number requested.

### CODE REVIEW

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxConnector.cpp [hunk 1]
```diff
@@ -1075,39 +1107,34 @@ HivePrestoToVeloxConnector::toVeloxSplit(
...
+      std::make_unique<velox::connector::hive::HiveConnectorSplit>(
+          catalogId,
+          hiveSplit->fileSplit.path,
+          toVeloxFileFormat(hiveSplit->storage.storageFormat),
+          hiveSplit->fileSplit.start,
+          hiveSplit->fileSplit.length,
+          partitionKeys,
+          hiveSplit->tableBucketNumber
+              ? std::optional<int>(*hiveSplit->tableBucketNumber)
+              : std::nullopt,
```
Comment (aditi-pandit) [inline presto-native-execution/presto_cpp/main/types/PrestoToVeloxConnector.cpp:1120]: Is this always null if there is bucketConversion ? Would be great to know if they are related.

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxConnector.cpp [hunk 2]
```diff
@@ -1028,6 +1028,38 @@ toHiveBucketProperty(
...
+      toHiveColumnType(hiveColumn->columnType),
+      stringToType(hiveColumn->typeSignature, typeParser),
+      hiveTypeParser.parse(hiveColumn->hiveType),
+      toRequiredSubfields(hiveColumn->requiredSubfields));
+}
+
+velox::connector::hive::HiveBucketConversion toVeloxBucketConversion(
+    const protocol::BucketConversion& bucketConversion) {
+  velox::connector::hive::HiveBucketConversion veloxBucketConversion;
+  veloxBucketConversion.tableBucketCount = bucketConversion.tableBucketCount;
```
Comment (aditi-pandit) [inline presto-native-execution/presto_cpp/main/types/PrestoToVeloxConnector.cpp:1052]: Would be great to know what these fields mean. Optionally : add documentation for them here in velox/connectors/hive/HiveConnectorSplit.h for https://github.com/facebookincubator/velox/pull/10228

**Review by aditi-pandit — COMMENTED** [review]
Thanks @Yuhta. Overall looks very good minus some documentation comments.

**Comment by aditi-pandit** [comment]
@Yuhta : We should have a Release note for this PR I feel. Please can you add details in the PR description.

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxConnector.cpp (hunk 1 above)
Comment (Yuhta) [reply presto-native-execution/presto_cpp/main/types/PrestoToVeloxConnector.cpp:1120]: It is always non-null if there is bucketConversion.  I have check in Velox but can add a check here as well

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxConnector.cpp [hunk 3]
```diff
@@ -1028,6 +1028,38 @@ toHiveBucketProperty(
       sortedBy);
 }
 
+std::unique_ptr<velox::connector::hive::HiveColumnHandle>
+toVeloxHiveColumnHandle(
+    const protocol::ColumnHandle* column,
+    const TypeParser& typeParser) {
+  auto hiveColumn = dynamic_cast<const protocol::HiveColumnHandle*>(column);
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/types/PrestoToVeloxConnector.cpp:1035]:
```
nit: s/auto/auto*/
```

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxConnector.cpp [hunk 4]
```diff
@@ -1075,39 +1107,34 @@ HivePrestoToVeloxConnector::toVeloxSplit(
...
-  velox::type::fbhive::HiveTypeParser hiveTypeParser;
-  // TODO(spershin): Should we pass something different than 'typeSignature'
-  // to 'hiveType' argument of the 'HiveColumnHandle' constructor?
-  return std::make_unique<velox::connector::hive::HiveColumnHandle>(
-      hiveColumn->name,
-      toHiveColumnType(hiveColumn->columnType),
-      stringToType(hiveColumn->typeSignature, typeParser),
-      hiveTypeParser.parse(hiveColumn->hiveType),
-      toRequiredSubfields(hiveColumn->requiredSubfields));
+  return toVeloxHiveColumnHandle(column, typeParser);
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/types/PrestoToVeloxConnector.cpp:1137]: There is no logic change to this function? Thanks!

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxConnector.cpp [hunk 5]
```diff
@@ -1028,6 +1028,38 @@ toHiveBucketProperty(
...
+}
+
+velox::connector::hive::HiveBucketConversion toVeloxBucketConversion(
+    const protocol::BucketConversion& bucketConversion) {
+  velox::connector::hive::HiveBucketConversion veloxBucketConversion;
+  veloxBucketConversion.tableBucketCount = bucketConversion.tableBucketCount;
+  veloxBucketConversion.partitionBucketCount =
+      bucketConversion.partitionBucketCount;
+  TypeParser typeParser;
+  for (auto& column : bucketConversion.bucketColumnHandles) {
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/types/PrestoToVeloxConnector.cpp:1056]:
```
nit: const auto& column
```

**Review by xiaoxmeng — APPROVED** [review]
@Yuhta thanks for the fix!

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxConnector.cpp (hunk 1 above)
Comment (aditi-pandit) [reply presto-native-execution/presto_cpp/main/types/PrestoToVeloxConnector.cpp:1120]: Would be great to add it here as well.

File: presto-native-execution/presto_cpp/main/types/tests/PrestoToVeloxSplitTest.cpp [hunk 6]
```diff
@@ -135,3 +135,32 @@ TEST_F(PrestoToVeloxSplitTest, serdeParameters) {
...
+  hiveSplit.bucketConversion = std::make_shared<protocol::BucketConversion>();
+  hiveSplit.bucketConversion->tableBucketCount = 4096;
+  hiveSplit.bucketConversion->partitionBucketCount = 512;
+  auto& column = hiveSplit.bucketConversion->bucketColumnHandles.emplace_back();
+  column.name = "c0";
+  column.hiveType = "bigint";
+  column.typeSignature = "bigint";
+  column.columnType = protocol::ColumnType::REGULAR;
+  auto veloxSplit = toVeloxSplit(scheduledSplit);
+  auto& veloxHiveSplit =
```
Comment (aditi-pandit) [inline presto-native-execution/presto_cpp/main/types/tests/PrestoToVeloxSplitTest.cpp:152]: Nit : const

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxConnector.cpp (hunk 4 above)
Comment (Yuhta) [reply presto-native-execution/presto_cpp/main/types/PrestoToVeloxConnector.cpp:1137]: No any logical change to existing code

**Review by aditi-pandit — DISMISSED** [review]
Thanks @Yuhta

**Review by xiaoxmeng — APPROVED** [review]

**Pull request merged**